casper/src/test/scala/coop/rchain/casper/batch2/BlockReceiverEffectsSpec.scala

```diff
@@ -0,0 +1,243 @@
+
+        val bsContainsCaptor = ArgCaptor[Seq[BlockHash]]
+        bs.contains(bsContainsCaptor) wasCalled 3.times
+        bsContainsCaptor.values should contain allOf (Seq(a1.blockHash), Seq(a2.blockHash))
+
+        val brAckReceivedCaptor = ArgCaptor[BlockHash]
+        br.ackReceived(brAckReceivedCaptor) wasCalled twice
+        brAckReceivedCaptor.values should contain allOf (a1.blockHash, a2.blockHash)
+
+        dagStorageWasNotModified(bds)
+        a1InOutQueue shouldBe a1.blockHash
+        a2InOutQueue shouldBe a2.blockHash
+      }
+  }
+
+  private def blockDagStorageMock[F[_]: Applicative](): BlockDagStorage[F] = {
+    val emptyDag = DagRepresentation(Set(), Map(), SortedMap(), DagMessageState(), Map())
+    mock[BlockDagStorage[F]].getRepresentation returnsF emptyDag
+  }
+
+  private def blockRetrieverMock[F[_]: Applicative](): BlockRetriever[F] =
+    mock[BlockRetriever[F]].ackReceived(*) returns ().pure[F]
+
+  private def blockStoreMock[F[_]: Sync: Applicative](): BlockStore[F] = {
+    val state  = Ref.unsafe[F, Map[BlockHash, BlockMessage]](Map())
+    val bsMock = mock[BlockStore[F]]
+    bsMock.contains(*) answers { keys: Seq[BlockHash] =>
+      state.get.map(s => Seq(s.contains(keys.head)))
+    }
+    bsMock.put(*) answers { kvPairs: Seq[(BlockHash, BlockMessage)] =>
+      state.update(s => kvPairs.foldLeft(s) { case (acc, item) => acc + item })
+    }
+    bsMock
+  }
+
+  import fs2._
+
+  private def withEnv(shardId: String)(
+      f: (
+          Queue[Task, BlockMessage],
+          Queue[Task, BlockMessage],
+          Stream[Task, BlockHash],
+          BlockStore[Task],
+          BlockRetriever[Task],
+          BlockDagStorage[Task]
+      ) => Task[Assertion]
+  ): Task[Assertion] =
+    for {
+      state                 <- Ref[Task].of(BlockReceiverState[BlockHash])
+      incomingBlockQueue    <- Queue.unbounded[Task, BlockMessage]
+      incomingBlockStream   = incomingBlockQueue.dequeue
+      validatedBlocksQueue  <- Queue.unbounded[Task, BlockMessage]
+      validatedBlocksStream = validatedBlocksQueue.dequeue
+
+      // Create mock separately for each test
+      bs  = blockStoreMock[Task]()
+      br  = blockRetrieverMock[Task]()
+      bds = blockDagStorageMock[Task]()
+
+      blockReceiver <- {
+        implicit val (bsImp, brImp, bdsImp) = (bs, br, bds)
+        BlockReceiver(state, incomingBlockStream, validatedBlocksStream, shardId)
+      }
+      res <- f(incomingBlockQueue, validatedBlocksQueue, blockReceiver, bs, br, bds)
+    } yield res
+
+  private def makeDefaultBlock =
+    BlockMessage
+      .from(
+        BlockMessageProto(
+          shardId = "root",
+          postStateHash = "abc".unsafeHexToByteString,
+          sigAlgorithm = Secp256k1.name
+        )
+      )
+      .right
+      .get
+
+  private def makeBlock(
+      justifications: List[BlockHash] = List()
+  ): Task[BlockMessage] = {
```

Comment:
This function doesn't have to be inside F. It's just regular Scala code, call to pure is unnecessary.